### PR TITLE
DEV-4784 | create a management command to fix imported contributions that incorrectly point to default donation page

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

This PR adds a management command to fix a subset of imported contributions created by Stripe transactions import.

There was a bug in the command implementation — fixed in [DEV-4783](https://news-revenue-hub.atlassian.net/browse/DEV-4783) — whereby default donation page for a contribution's RP got assigned as the value for contribution.donation_page, even when this wasn't warranted.

The management command herein identifies contributions that ended up in this state.  Specifically, it looks for contributions that:

- Have a revision comment indicating they were created by Stripe transactions import
- Have contribution metadata that either doesn't have an entry for `referer` or an empty string for that value
- Have a donation page set, and that donation page is the same donation page as the parent RP's default donation page

For these contributions, the command deletes the donation_page foreign key and sets `.revenue_program` to the revenue program indicated by the `revenue_program_id` field in the metadata.

Additionally, this PR makes a small change to `StripeTransactionsImporter. get_or_create_contributor_from_customer` to solve for an edge case discovered when running transactions import locally when a Stripe customer is encountered that has no value for the `email` property.

#### Why are we doing this? How does it help us?

Retroactively solve for contributions created when Stripe transactions import command had a bug around incorrectly assigning default donation page.

Left unchecked, this bug would potentially prevent orgs from deleting a donation page that in actuality has had no contributions generated by it, but where contributions were linked only because of the bug (when they should have had `._revenue_program` assigned).

#### Are there detailed, specific, step-by-step testing instructions in the associated ticket?

Yes.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

No

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-4784](https://news-revenue-hub.atlassian.net/browse/DEV-4784)
- [DEV-4783](https://news-revenue-hub.atlassian.net/browse/DEV-4783)
- [ITS-2375](https://news-revenue-hub.atlassian.net/browse/ITS-2375) 
- [DEV-4788](https://news-revenue-hub.atlassian.net/browse/DEV-4788)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)? If so, open a ticket and link it to the ticket for this PR and list it here:

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

- [ITS-2375](https://news-revenue-hub.atlassian.net/browse/ITS-2375) to run this command in prod after this PR has been merged.
- [DEV-4788](https://news-revenue-hub.atlassian.net/browse/DEV-4788), which is to retire this code once ITS-2375 has been done.

[DEV-4783]: https://news-revenue-hub.atlassian.net/browse/DEV-4783?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DEV-4784]: https://news-revenue-hub.atlassian.net/browse/DEV-4784?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+from django.db.models.functions import Cast
+
+import reversion
+from reversion.models import Revision
+
+from apps.contributions.models import Contribution
+from apps.organizations.models import RevenueProgram
+
+
+REVISION_COMMENT = "StripeTransactionsImporter.upsert_contribution created Contribution"
```
Comment (njharman) [inline apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py:12]: I believe this string is derived from utils.py::upsert_with_diff_check() There are two paths in that function, "created" and "updated". 

Checking, asking; updated Contributions can't or didn't create any of the bad records being corrected with this command?

File: apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+class Command(BaseCommand):
+    """Find recurring contributions affected by bug solved for in DEV-4783 and fix them by...
+
+    ...setting donation_page to None and setting `._revenue_program` to the correct RevenueProgram.
+    """
+
+    def handle(self, *args, **options):
+        self.stdout.write(
+            self.style.HTTP_INFO("Running `fix_imported_contributions_with_incorrect_donation_page_value`")
```
Comment (njharman) [inline apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py:23]: To protect against filename changes/DRY. I tend to make use of __file__ for these kind of messages. Maybe you have to strip off path, I think so. Path.stem()

**Review by njharman — APPROVED** [review]

File: apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py (hunk 1 above)
Comment (benjaminEwhite) [reply apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py:12]: That is correct. We only want to target ones that were created, not ones that were updated. In fact, updated ones should not have been able to be put in the state caused by original bug necessitating this management command.

File: apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py (hunk 2 above)
Comment (benjaminEwhite) [reply apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py]: d95b4bdfc4

File: apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            return
+
+        self.stdout.write(
+            self.style.HTTP_INFO(
+                f"Found {contributions} eligible contribution{'' if contributions == 1 else 's'} to fix"
+            )
+        )
+        updated_ids = []
+        for x in contributions.all():
+            rp = RevenueProgram.objects.get(pk=x.contribution_metadata["revenue_program_id"])
```
Comment (x110dc) [inline apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py:58]: minor: looks like `rp` variable is only used once so probably didn't need it

File: apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            ~Q(contribution_metadata__has_key="referer") | Q(contribution_metadata__referer=""),
+            ~Q(contribution_metadata__revenue_program_id=""),
+            contribution_metadata__has_key="revenue_program_id",
+            donation_page__isnull=False,
+            # donation is set to same value as default donation page of the rp
+            donation_page=F("donation_page__revenue_program__default_donation_page"),
+        )
+
+        if not contributions.exists():
+            self.stdout.write(self.style.HTTP_INFO("No effected contributions found, exiting"))
```
Comment (x110dc) [inline apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py:48]: super minor: affected, not effected

File: apps/contributions/stripe_import.py [hunk 5]
```diff
@@ -612,14 +614,14 @@ def get_payment_method(self, pm_id: str) -> stripe.PaymentMethod:
         """Get a payment method from stripe"""
         return stripe.PaymentMethod.retrieve(pm_id, stripe_account=self.stripe_account_id)
 
-    def get_payment_method_for_stripe_entity(
+    def get_payment_method_id_for_stripe_entity(
```
Comment (x110dc) [inline apps/contributions/stripe_import.py:617]: the set of changes in this file doesn't seem to be mentioned in the PR description? Maybe I'm misunderstanding.

File: apps/contributions/stripe_import.py [hunk 6]
```diff
@@ -872,15 +875,16 @@ def process_transactions_for_recurring_contributions(self) -> None:
...
-        without referer and schema_version, we know ahead of time that all of the PIs we're looking at are for one-time contributions.
-        """
+    def process_transactions_for_one_time_contributions(self) -> None:
+        """Process transactions for one-time contributions."""
         logger.info("Processing transactions for one-time contributions")
-        for key in self.redis.scan_iter(match=self.make_key(entity_name="PaymentIntent_*")):
-            pi = self.get_resource_from_cache(key)
+        for pi in self.payment_intents:
             try:
                 contribution, action = self.upsert_contribution(stripe_entity=pi, is_one_time=True)
```
Comment (x110dc) [inline apps/contributions/stripe_import.py:889]: The PR description says: 

> Additionally, this PR makes a small change to StripeTransactionsImporter. get_or_create_contributor_from_customer to solve for an edge case discovered when running transactions import locally when a Stripe customer is encountered that has no value for the email property.

But I don't see those changes in this PR.

**Review by x110dc — CHANGES_REQUESTED** [review]
a couple very minor changes and some possible discrepancies with what's in the this PR and what's in the PR description

File: apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py (hunk 4 above)
Comment (benjaminEwhite) [reply apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py:48]: 145033e265dc

File: apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py (hunk 3 above)
Comment (benjaminEwhite) [reply apps/contributions/management/commands/fix_imported_contributions_with_incorrect_donation_page_value.py]: b4a8d52a01

File: apps/contributions/stripe_import.py (hunk 5 above)
Comment (benjaminEwhite) [reply apps/contributions/stripe_import.py]: Not misunderstanding. I seem to have accidentally committed/pushed changes from another branch when switching between branches locally. I've removed the errant changes, so everything should be in line with PR now.

File: apps/contributions/stripe_import.py (hunk 6 above)
Comment (benjaminEwhite) [reply apps/contributions/stripe_import.py]: See previous comment. I seem to have accidentally saved some changes from a different branch into this one, likely when switching between branches with VSCode open.

Changes should now align with PR description.

**Review by x110dc — APPROVED** [review]
LGTM.

**Pull request merged**